Computational Geometry Refactor

srwopschall opened the pull request:
This PR consists of a computational geometry (CG) refactor that separates CG logic and routines on a per contact method basis. This branch came about when trying to remove gap separation logic in favor of binning proximity, where bugs were encountered that erroneously switched common plane overlap calculations from interpen to full overlap. This behavior had the potential to lead to unstable face-pair interaction behavior. In general, this PR does the following:

- separate common plane, mortar, and aligned mortar methods' CG logic and routines under a shared ContactPlanePair base class
- Coalesce the CG routines in a common interface/class-hierarchy allowing for easier method based extensions in the future.
- Create a common check based structure to the CG code path, which in turn simplifies testing and debugging and future implementations.
- Various bug fixes along the way; easier handling of edge cases with refactored routines.
- Moved/re-implemented a finer proximity check to geomFilter() to avoid unnecessary memory allocation, specifically an issue on GPUs. This also simplified the list of checks later on.
- Added numerous 2D common plane and 3D common plane and single mortar tests. These test very specific and potentially difficult face-pair configurations that not only test the CG routines but the whole code path. Tests compare to exact geometric solutions.
- Added a common plane test using previous host-code face-pair coordinates that once resulted in an inverted overlapping polygon; refactor fixed this issue.

srwopschall committed:
Initial plumbing to remove gap separation tolerance in preference of binning proximity.

srwopschall committed:
Convert FaceInterCheck() to a check for full overlap calc.

srwopschall committed:
Modified EdgeInterCheck() to check for full overlap trigger.

srwopschall committed:
Fixed compiler error.

srwopschall committed:
Formatting and removed negative tol in favor of direct comparison to
zero for overlap logic

srwopschall committed:
Add logic for interpenOverlap calc to switch to fullOverlap for detected
configurations; Face/EdgeInterCheck() was not doing this correctly

srwopschall committed:
Minor comments and new face centroid function.

srwopschall committed:
renamed ContactPlane.* to CompGeom.* for generality

srwopschall committed:
Renamed contact plane files to CompGeom.*; updated headers and cmake; plumbing for reorganizing the class interface; did not update .cpp yet.

srwopschall committed:
More plumbing for CG classes; moved free function from CompGeom to
GeomUtilities

srwopschall committed:
Refining the comp geom classes and constructors. Moving away from
contact planes on the coupling scheme.

srwopschall committed:
Getters on the CompGeom object to return individual contact planes.

srwopschall committed:
Modified CheckInterfacePair() to work with new CompGeom container.

srwopschall committed:
Updated comp geom checkInterfacePair() function signature

srwopschall committed:
Updated comp geom class constructors and removed old 2D/3D contact plane
constructors

srwopschall committed:
Plumbing for checkInterfacePair() wrappers on method specific check
routines.

srwopschall committed:
Further plumbing for all checkInterfacePair() routines

srwopschall committed:
Overhauled Common Plane 3D intersection routines; added point in face
check in geomFilter(); plumbing for mortar 3D face check routines.

srwopschall committed:
Fixed common plane intersection routine to handle full overlap switch;
added/use helper routines to project 3D planar points to 2D.

srwopschall committed:
Fixed doxygen note for 3d point to 2d projection helper routine

srwopschall committed:
Further fixed doxygen for new helper routine

srwopschall committed:
Modified interpen vs. full overlap logic; default full overlap calc is
true for all face based methods.

srwopschall committed:
Fixed syntax on bool initialization

srwopschall committed:
MortarPlanePair computeOverlap3D() implementation

srwopschall committed:
Cleanup on who sets the inContact boolean

srwopschall committed:
Moved computeAreaTol() to ContactPlanePair and added min area check for
mortar face check.

srwopschall committed:
Refactored aligned mortar cg methods; refactored each plane type's
normal and centroid calc.

srwopschall committed:
Refactored 2D checkEdgePair; made CheckSegOverlap() a free function;
removed/modified old ContactPlane2D/3D routines; started geomFilter() 2D
overlap proxy check.

srwopschall committed:
Consolidated projection and overlap calc for common plane overlap in new
routine; allows to switch from interpen to full overlap calc in edge
case.

srwopschall committed:
Added 2D overlap proxy check #5 in geomfilter()

srwopschall committed:
Fixed all cpu compiler errors and most warnings.

srwopschall committed:
Fixed bug in allocation of comp geom pairs arrays.

srwopschall committed:
added doxygen for new pairs allocation routine on the comp geom class.

srwopschall committed:
Fixed indexing issue in various projections.

srwopschall committed:
Fixed check #5 using point in face queries as proxy for overlaps

srwopschall committed:
Moving to cheaper overlap area calc for check 5; point in face marks
adjacent faces as being in contact which overcounts/allocates

srwopschall committed:
Minor comment formatting.

srwopschall committed:
Separating 3d comp geom tests from 2d in new files.

srwopschall committed:
Added edge case for switch from full to interpen; added various tests
cases; debugged projections used in common plane gap calculation.

srwopschall committed:
Reworked setting area centroid and resetting contact plane; removed
local overlap centroid data (unused); added gap diff check to tests.

srwopschall committed:
added second interpen check with exact overlap and gap checks for
triangular interpen face; will remove print statements later.

srwopschall committed:
Added more 3D tests for interpen overlap calcs.

srwopschall committed:
Fixed a 3D bug in Intersection2DPolygon; Added another complicated
intersection test

srwopschall committed:
Added another common plane test with max overlap polygon vertices; added
shift feature to test mesh rotation.

srwopschall committed:
Removed some print statements

srwopschall committed:
Added 3D test with face geom from host code that used to produce
inverted overlap; bug is fixed.

srwopschall committed:
removed old point in face check in geomfilter.

srwopschall committed:
Added 2d common plane tests.

srwopschall commented:
Finished with adding specific 2D and 3D computational geometry tests so placing this up as a draft PR.

TODO:
- There are two mortar tests that use host-code surface coordinates that are failing for the precise reason the tests were created in the first place. Namely, the inverse isoparametric mapping is finding points that lie outside the parent quadrilateral. This is the last thing I have to debug. I suspect this is due to a face-pair edge case that needs to be handled in the new refactored CG framework.
- Ensure the GPU build and the Enzyme tests pass. Testing/debugging has primarily been focused on the CPU and non-enzyme functionality.
- Verify existing tests in case results comparisons need to be updated due to different CG results with refactor/bugfixes.

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
Merge branch 'develop' into feature/swopscha/remove_separation_check

srwopschall committed:
Pushed proper coordinate storage and use through the physics routines;
added helper/getter functions

srwopschall commented:
Update: the CG issue with InvIso has been resolved. All tests pass except for the multi-rank mfem common plane tests. Will meet with @ebchin to see what might cause this since the serial tests passes, the CG is working, and the CG would not have affected parallel runs.

srwopschall committed:
Removed resetting binning proximity; mfem common plane with k=2 were failing

srwopschall committed:
Fixed bad allocation size in PolyReorder() for the proj array; this
caused mfem common plane for polynomial order 1 to fail.

srwopschall committed:
Increase max size for allocation in PolyReorderWithNormal(); max face
for overlap size can be 5.

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
Fixed bug when switching from interpen to full overlap calc.

srwopschall committed:
merge formatting

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
Removed unnecessary and unguarded slic info

srwopschall committed:
Fixed cg view getter in enzyme code.

srwopschall committed:
Fixing comp geom view getter in enzyme mortar. Guarded slic call.

srwopschall committed:
Removed host device decoration where not needed.

srwopschall committed:
Removed host device decoration on CompGeom constructor

srwopschall committed:
Test removing slic calls from coupling scheme apply.

srwopschall committed:
Restoring slic calls.

srwopschall committed:
Test removing call to writing vtk output.

srwopschall committed:
Testing mpi include

srwopschall commented:
/style

srwopschall committed:
Removed mpi include that was done to test build.

srwopschall commented:
Putting this up for review. There is a very strange compiler error just on blueos that I'm still hunting down, but everything else passes CI.

srwopschall committed:
Removed slic header in comp geom.

white238 commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1975 @@
+  // These are the 'prime' coordinates and ensure that our cg is working on
+  // truly planar 4 node quadrilaterals
+
+  // get face normmals
```
```suggestion
  // get face normals
```

srwopschall committed:
Removed getter to return ArrayViewT<> of contact pairs; updated mortar
and aligned mortar.

srwopschall committed:
Fixed compiler error

srwopschall commented:
/style

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
Removed unnecessary header, removed duplicate files in cmake list

srwopschall committed:
Decorating geom utilities function with host device.

srwopschall committed:
Fixed rocm build by adding missing host config; relaxed gpu tolerance (diffs in the noise)

srwopschall committed:
adding rzvernal host config

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
Fixed bug in assigning number of 2D overlap vertices

srwopschall committed:
Merge branch 'feature/swopscha/remove_separation_check' of github.com:LLNL/Tribol into feature/swopscha/remove_separation_check

srwopschall committed:
Moved problem communicator from parameters to coupling scheme.

srwopschall committed:
Add mpi header

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
Fixed loops to store global projected overlapping faces in 3D.

srwopschall committed:
Fixed bug in specifying max number of overlap vertices.

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
Changed 8 max overlap verts to 10 for consistency

srwopschall committed:
Merge branch 'feature/swopscha/remove_separation_check' of github.com:LLNL/Tribol into feature/swopscha/remove_separation_check

srwopschall committed:
Removed TODO and cout statements

srwopschall committed:
Release notes.

srwopschall committed:
Added 2D common plane penalty test with force check; removed prints.

srwopschall committed:
Removed more prints; changed force sense comparison to ensure
approximately zero nodal force is caught as error.

srwopschall commented:
/style

Commit:
Apply style updates

srwopschall committed:
cout cleanup

srwopschall commented:
This PR is ready for review. The fmt/slic issue affecting the blueos build has been fixed. CI has been run separately on cray and passes (I'm awaiting access to the required machine so I can launch these jobs), and any extraneous formatting fail(s) will be fixed prior to future merge. Also, this branch has been tested/debugged in appropriate host codes and test behavior has been verified.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -1313,8 +1571,9 @@ TRIBOL_HOST_DEVICE bool PolyReorder( RealT* x, RealT* y, int* newIDs, int numPoi
-  RealT proj[max_nodes_per_overlap - 2];
+  constexpr int max_nodes_per_overlap = 10 + 2 * 4;
+  constexpr int max_nodes = 4;
+  RealT proj[max_nodes];
```
Looks like this needs to be sized to `max_nodes_per_overlap - 2`, or something larger than 4.

srwopschall replied:
Yes. I set this to 8. Correct behavior confirmed.

srwopschall committed:
Fix bug in max overlap projections in PolyReorder

ebchin commented on src/tests/tribol_common_plane_gpu.cpp:
```diff
@@ -64,7 +64,7 @@ class CommonPlaneGPUTest : public testing::TestWithParam<ExecutionMode> {
 TEST_P( CommonPlaneGPUTest, update_test )
 {
-  EXPECT_LT( max_error_, 1.0e-12 );
+  EXPECT_LT( std::abs( max_error_ ), 1.0e-10 );
```
Good catch.  Any idea why the error went up?

srwopschall replied:
I can't say with confidence. I played around with the mref and the penalty and couldn't seem to get the error below 1.e-10. I think it is fine for now in the sense that 1.e-10 still ought to catch any introduction of a bug or bad modification to the algorithm.

ebchin commented on src/tribol/mesh/CouplingScheme.cpp:
```diff
@@ -14,7 +14,7 @@
 #include "tribol/search/InterfacePairFinder.hpp"
 #include "tribol/common/Parameters.hpp"
-#include "tribol/geom/ContactPlane.hpp"
+#include "tribol/geom/CompGeom.hpp"
```
Consider removing, since this is already `#include`d in the header.

ebchin replied:
More generally, it might be worth it to do a header cleanup PR to remove unused headers.  clang-tidy can help with this.  Could potentially improve compile times.

srwopschall replied:
Removed the include. I agree. Going through this PR I realize the headers do need a cleanup. Follow-on PR.

ebchin commented on src/tribol/interface/tribol.cpp:
```diff
@@ -12,7 +12,7 @@
 #include "tribol/mesh/InterfacePairs.hpp"
 
-#include "tribol/geom/ContactPlane.hpp"
+#include "tribol/geom/CompGeom.hpp"
```
Is anything in `tribol.cpp` using something in `CompGeom.hpp` directly?  Maybe we can remove this include.

srwopschall replied:
Removed.

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -153,12 +154,18 @@ void checkForceSense( tribol::CouplingScheme const* cs, bool isTied = false )
         if ( !isTied ) {
           // <= catches interpenetration AND separation
-          EXPECT_LE( force_mag, 0. );
+          EXPECT_LE( force_mag, -1.e-6 );
```
Why did this need to be changed?

srwopschall replied:
I don't recall. Just changed it back and the test passes.

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -501,6 +508,94 @@ TEST_F( CommonPlaneTest, tied_contact_check )
+  RealT gap = -2. * half_gap;
+  compareGaps( couplingScheme, gap, 1.E-8, "kinematic_penetration" );
+}
+
```
Nice added check

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -501,6 +508,94 @@ TEST_F( CommonPlaneTest, tied_contact_check )
+    y1[i] = xy1[i * dim + 1];
+    x2[i] = xy2[i * dim] + x_shift;
+    y2[i] = xy2[i * dim + 1];
+  }
```
Can this be cleaned up to be defined once?  The re-definition threw me off as to what the coordinates were.

srwopschall replied:
Good call. Fixed.

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -501,6 +508,94 @@ TEST_F( CommonPlaneTest, tied_contact_check )
+
+  checkForceSense( couplingScheme );
+
+  RealT length = 0.75 - ( xy2[0] + x_shift );
```
These calcs aren't clear to me, shouldn't the length and height just be the diff of the x and y coords?

srwopschall replied:
I added clarifying comments where appropriate as best I could. It is hard to fully communicate what the calc is doing without drawing it. The new comments at least act as a bit of a guide. I removed hard coded numbers as seen here for clarity.

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -501,6 +508,94 @@ TEST_F( CommonPlaneTest, tied_contact_check )
+  RealT computed_area = hyp * std::cos( half_theta );
+  RealT half_area = 0.5 * computed_area;
+  RealT half_gap = half_area * std::tan( half_theta );
+  RealT gap = -2. * half_gap;
```
This is the gap calc at the centroid of the overlap?  Can this be simplified?

srwopschall replied:
Easiest way to simplify is just hard code the number, but I like having the analytic geometric calc in here. I agree it is a little hard to parse. I'd prefer to leave this the way it is (and in the 2D and 3D comp geom tests).

ebchin commented on src/tests/tribol_comp_geom_3d.cpp:
```diff
@@ -0,0 +1,1652 @@
+  }
+
+ protected:
+};
```
Let's add some tests for warped elements

ebchin replied:
Does it make sense to do these tests again, only with mortar and with different expected results, just to make sure mortar works under these scenarios?  Maybe we can turn the test into a parametric test?

srwopschall replied:
I'll add some test(s) for warped elements. I have two single mortar checks, one for an x-like configuration and one for a full penetration configuration, and I have one aligned mortar check. I think this is fine for now. I can add a warped test for each method, which may help test coverage.

srwopschall replied:
Added a common plane warped test.

ebchin commented on src/tests/tribol_common_plane_gap_rate.cpp:
```diff
@@ -41,7 +42,7 @@ void compareGaps( tribol::CouplingScheme const* cs, RealT gap, const RealT tol,
 
   for ( tribol::IndexT cpID = 0; cpID < numPairs; ++cpID ) {
-    auto& plane = cs->getContactPlane( cpID );
+    auto& plane = static_cast<const tribol::CommonPlanePair&>( cs->getContactPlanePair( cpID ) );
```
Would `getCompGeom().getCommonPlane( cpID );` work instead?

srwopschall replied:
yes, this works and is cleaner. I updated everywhere.

ebchin commented on src/tests/tribol_common_plane_gap_rate.cpp:
```diff
@@ -116,7 +117,7 @@ void checkPressures( tribol::CouplingScheme const* cs, RealT pressure, const Rea
 
   for ( tribol::IndexT cpID = 0; cpID < numPairs; ++cpID ) {
-    auto& plane = cs->getContactPlane( cpID );
+    auto& plane = static_cast<const tribol::CommonPlanePair&>( cs->getContactPlanePair( cpID ) );
```
Would `getCompGeom().getCommonPlane( cpID );` work instead?

srwopschall replied:
Updated.

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -42,7 +43,7 @@ void compareGaps( tribol::CouplingScheme const* cs, RealT gap, const RealT tol,
 
   for ( tribol::IndexT cpID = 0; cpID < numPairs; ++cpID ) {
-    auto& plane = cs->getContactPlane( cpID );
+    auto& plane = static_cast<const tribol::CommonPlanePair&>( cs->getContactPlanePair( cpID ) );
```
Would `getCompGeom().getCommonPlane( cpID );` work instead?

srwopschall replied:
Updated.

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -116,7 +117,7 @@ void checkPressures( tribol::CouplingScheme const* cs, RealT pressure, const Rea
 
   for ( tribol::IndexT cpID = 0; cpID < numPairs; ++cpID ) {
-    auto& plane = cs->getContactPlane( cpID );
+    auto& plane = static_cast<const tribol::CommonPlanePair&>( cs->getContactPlanePair( cpID ) );
```
Would `getCompGeom().getCommonPlane( cpID );` work instead?

srwopschall replied:
Updated.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -57,238 +58,244 @@ class CompGeomTest : public ::testing::Test {
 
-  tribol::finalize();
-}
+  for ( int i = 0; i < numVerts; ++i ) {
```
Can this be cleaned up?  Maybe initialize directly into x1, etc. or send xy1 and xy2 directly into the API calls?

ebchin replied:
Same for the remaining tests

srwopschall replied:
These were all cleaned up by initializing directly into x1, etc. for all tests.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -57,238 +58,244 @@ class CompGeomTest : public ::testing::Test {
-  EXPECT_LE( diff_mag, tol );
+  EXPECT_EQ( 1, couplingScheme->getNumActivePairs() );
+
+  auto& plane = static_cast<const tribol::CommonPlanePair&>( couplingScheme->getContactPlanePair( 0 ) );
```
Get rid of the static_cast if possible (using `getCompGeom().getCommonPlane()`).  Check elsewhere in the PR too.

srwopschall replied:
Updated.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -57,238 +58,244 @@ class CompGeomTest : public ::testing::Test {
+  EXPECT_EQ( 1, couplingScheme->getNumActivePairs() );
+
+  auto& plane = static_cast<const tribol::CommonPlanePair&>( couplingScheme->getContactPlanePair( 0 ) );
+  RealT diff = std::abs( ( xy2[2] - xy2[0] ) - plane.m_area );
```
The tests mentions the vertices are coincident (to a tolerance, I assume?).  Can you add a comment why the coords in `xy2` are appropriate for this check?

srwopschall replied:
Added clarifying comments

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -57,238 +58,244 @@ class CompGeomTest : public ::testing::Test {
-  xy1[3] = 0.722646;
+  // this geometry has a pair of 'nearly' coincident vertices that should
+  // produce NO positive area of overlap. Note: the actual overlap is
+  // less than the contact area fraction set by tribol::setContactAreaFrac() below.
```
Is the value in `setContactAreaFrac()` scaled by dimension?  And is the area computed on the parent or physical coordinates?

srwopschall replied:
It is the fraction of the largest current configuration face area in a face-pair that provides the minimum overlap area threshold.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -480,29 +495,134 @@ TEST_F( CompGeomTest, coincident_vertex_no_overlap )
+  //    --------o----
+  //          *
+  //        *
+  //
```
Thanks, the illustration is helpful

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -480,29 +495,134 @@ TEST_F( CompGeomTest, coincident_vertex_no_overlap )
+  auto& plane = static_cast<const tribol::CommonPlanePair&>( couplingScheme->getContactPlanePair( 0 ) );
+
+  EXPECT_EQ( plane.m_fullOverlap, false );
+  RealT length = xy1[0] - xy1[2];
```
I think it would be clearer to do the inverse; i.e. define the length and height at the beginning of the test, then compute element coordinates based on that.

srwopschall replied:
I'll leave it for now. For me it is easier to envision the actual 2D coords of the edges and specify those first, rather than sort of parameterizing them based on a height and length, which makes how the faces interact a little more abstract. Perhaps I can add comments on these geometric calculations to get analytic overlap area and gap to make things clearer?

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -480,29 +495,134 @@ TEST_F( CompGeomTest, coincident_vertex_no_overlap )
+  RealT theta = std::atan( height / length );
+  RealT half_theta = 0.5 * theta;
+  RealT computed_area = 0.5 * length * std::cos( half_theta );
+  RealT diff = std::abs( computed_area - plane.m_area );
```
I've seen this in a few places now, would it make sense to put it in a function that describes what it's doing?  Or consider simplifying the code if it isn't needed.

srwopschall replied:
I think this only appears in two locations. I'll leave for now, but look into if I can consolidate or standardize some of the geometric calculations.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -480,29 +495,134 @@ TEST_F( CompGeomTest, coincident_vertex_no_overlap )
+  tribol::IndexT conn1[2] = { 0, 1 };
+  tribol::IndexT conn2[2] = { 0, 1 };
+
+  tribol::registerMesh( 0, 1, 2, &conn1[0], (int)( tribol::LINEAR_EDGE ), &x1[0], &y1[0], nullptr,
```
Can we put this block of code in a function for the test?  It might clean things up.

srwopschall replied:
I put all these register calls and the update call into a helper function on the test class. Cleans things up a bit.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -480,29 +495,134 @@ TEST_F( CompGeomTest, coincident_vertex_no_overlap )
+  //    ------------o
+  //             *
+  //          *
+  //
```
👍

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -616,10 +741,10 @@ TEST_F( CompGeomTest, 2d_projections_1 )
   RealT diffy1 = std::abs( cxProj1[1] - 0.0915028 );
```
This looks like an old test.  Do you know where the coordinates came from and what it's trying to capture?

srwopschall replied:
It is an old test. I don't recall where they came from but they are pretty specific I wouldn't be surprised if they originated in early testing and were copied over as a unit test. They are full overlap with two perfectly coincident nodes.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -659,12 +784,16 @@ TEST_F( CompGeomTest, 2d_projections_1 )
+  EXPECT_EQ( 1, couplingScheme->getNumActivePairs() );
 }
 
 TEST_F( CompGeomTest, 2d_projections_2 )
```
Similar for this test, what are the coordinates trying to test?

srwopschall replied:
A once problematic configuration that arose in early day testing.

ebchin commented on src/tests/tribol_comp_geom_3d.cpp:
```diff
@@ -0,0 +1,1652 @@
+  int nNonmortarElems = 1;
+  int nElemsXS = nNonmortarElems;
+  int nElemsYS = 1;
+  int nElemsZS = nNonmortarElems;
```
Does the `Mortar`/`Nonmortar` naming need to be updated for the `CommonPlane` tests?

srwopschall replied:
I modified all the 3d common plane tests to not use this language.

ebchin commented on src/tests/tribol_comp_geom_3d.cpp:
```diff
@@ -0,0 +1,1652 @@
+  RealT area_diff = std::abs( plane.m_area - 0.5 );
+  EXPECT_LE( area_diff, 1.e-10 );
+
+  tribol::finalize();
```
I think a function might clean up some of the duplication in these tests

srwopschall replied:
I introduced a math function to take the absolute value diffs rather than having a specific "area check" or "gap check" function. Sort of simplifies things. Embedded the call inside the EXPECT_LE. Does remove some code. Didn't modify all locations an absolute value diff is taken as that is a little beyond the scope of the PR, but did clean up the new comp geom tests.

ebchin commented on src/tribol/utils/TestUtils.cpp:
```diff
@@ -686,6 +686,100 @@ void TestMesh::setupContactMeshTet( int numElemsX1, int numElemsY1, int numElems
   this->mesh_constructed = true;
 
 }  // end setupContactMeshTet()
+//------------------------------------------------------------------------------
```
We should reconcile `TestMesh` with `MeshBuilder` at some point.  We could add this function to the `MeshBuilder` class, then have `TestMesh` own a `MeshBuilder` object as an interim solution.

ebchin commented on src/tests/tribol_comp_geom_3d.cpp:
```diff
@@ -0,0 +1,1652 @@
+TEST_F( CompGeomTest, auto_contact_lt_max_interpen )
+{
+  // This test uses auto-contact and checks that the face-pair
+  // is included as a conatct candidate, and is in fact in contact
```
contact

srwopschall replied:
fixed.

ebchin commented on src/tests/tribol_timestep_vote.cpp:
```diff
@@ -703,20 +703,20 @@ TEST_F( CommonPlaneTest, large_velocity_large_separation )
   RealT velX1 = 0.;
   RealT velY1 = 0.;
-  RealT velZ1 = vel_factor * 0.3 * element_thickness1 / dt;
+  RealT velZ1 = 0.5 * vel_factor * ( z_min2 - z_max1 + 0.25 * element_thickness1 ) / dt;
```
Are these changes bugfixes?

srwopschall replied:
This is a bugfix. It looks like the velZ2 had this change, but velZ1 didn't. The two blocks are moving toward one another hence the 0.5. Then, the distance of interpen is computed as the gap of separation between the two (z_min2-z_max1) and the amount of projected interpen is 0.25 the element thickness. This is consistent with the comment, saying that it is less than the 0.30 x element_thickness that would trigger a timestep vote.

ebchin commented on src/tribol/common/Parameters.hpp:
```diff
@@ -311,12 +311,14 @@ enum BasisEvalType
-  INVALID_FACE_INPUT,                          ///! Invalid input
-  DEGENERATE_OVERLAP,                          ///! Issues with overlap calculation resulting in degenerate overlap
+  NO_FACE_GEOM_ERROR,                 ///! No face geometry error indicating an overlap intersection
+  NO_OVERLAP,                         ///! Indicates that no actual overlap exists, but does not indicate an error
```
Should we create a `FaceGeomInfo` struct for this?  Similar to `CaseInfo` and `EnforcementInfo`?

ebchin replied:
Maybe we just report `NO_FACE_GEOM_ERROR` when there is no overlap?

srwopschall replied:
That is basically the idea...is that the only thing you care about is if there is an overlap or no overlap with no errors. I think a FaceGeomInfo would be more consistent, but would potentially introduce more logic/checks because I need to report an error and/or info. I'll leave as is for now because all the checks work nicely with checking for NO_FACE_GEOM_ERROR or != NO_FACE_GEOM_ERROR

ebchin commented on src/tribol/common/Parameters.hpp:
```diff
@@ -311,12 +311,14 @@ enum BasisEvalType
+  FACE_ORIENTATION,                   ///! Face vertices not ordered consistent with outward unit normal
+  INVALID_FACE_INPUT,                 ///! Invalid input
+  DEGENERATE_OVERLAP,                 ///! Issues with overlap calculation resulting in degenerate overlap
+  EXCEEDS_AUTO_CONTACT_LENGTH_SCALE,  ///! Gap exceeds auto contact length scale
```
Same with this one, I don't think this is an error case?

srwopschall replied:
I changed the enum to FaceGeomException. I think this better captures how the enum is being used.

ebchin commented on src/tribol/geom/CompGeom.hpp:
```diff
@@ -0,0 +1,1021 @@
+   *
+   * \return face geometry error
+   */
+  TRIBOL_HOST_DEVICE virtual FaceGeomError checkInterfacePair( const MeshData::Viewer& mesh1,
```
Is this virtual function needed?

srwopschall replied:
Per this design, yes. Is it strictly necessary? Maybe not, but inside this virtual function is where a method specific switch on dimension is embedded. If this was higher up, that switch would clutter up a higher level function. Additionally, there may be additional things someone wants to do when checking for a method specific interface pair, which could be planed in this virtual functions implementation. I'll keep for now.

ebchin replied:
The comment is regarding the keyword virtual.  I don't think it is needed and indicates the function is used polymorphically, which I don't think is the case.

srwopschall replied:
I see. I think I misunderstood. I will rework it.

ebchin commented on src/tribol/geom/CompGeom.hpp:
```diff
@@ -0,0 +1,1021 @@
+   *
+   * \return face geometry error
+   */
+  TRIBOL_HOST_DEVICE virtual FaceGeomError checkFacePair( const MeshData::Viewer& mesh1,
```
Same with this one and the following ones.  I don't think we will ever be using polymorphism on a ContactPlanePair.

srwopschall replied:
Per this design, each method specific computational geometry class has to implement their own version of checkFacePair(). Mortar, AlignedMortar, and Common Plane all have different logic. I opted for this rather than having the classes have their own `checkCommonPlaneFacePair()`, for instance. Additionally, `checkFacePair()` is not simply a routine on the `ContactPlanePair` base class because that would put us right back where we started with a very convoluted set of method specific checks embedded within one another. I will keep the comment unresolved so we can discuss, but for now I'll keep this; I think it is clean.

srwopschall replied:
I'll rework this.

ebchin commented on src/tribol/geom/CompGeom.hpp:
```diff
@@ -0,0 +1,1021 @@
+   * \pre length(pX), length(pY), length(pZ) >= size
+   * \pre length(pLX), length(pLY) >= size
+   */
+  TRIBOL_HOST_DEVICE void globalTo2DLocalCoords( RealT* pX, RealT* pY, RealT* pZ, RealT* pLX, RealT* pLY, int size );
```
```suggestion
  TRIBOL_HOST_DEVICE void globalTo2DLocalCoords( const RealT* pX, const RealT* pY, const RealT* pZ, RealT* pLX, RealT* pLY, int size );
```

srwopschall replied:
Good catch. Fixed.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -20,6 +21,87 @@
+  // if the point (x,y,z) is really close to the centroid
+  RealT sqrMag = e1x * e1x + e1y * e1y + e1z * e1z;
+  if ( sqrMag < 1.E-12 ) {
+    RealT scale = 10.;  // just scale it 10x the original distance
```
```suggestion
    RealT scale = 10.;  // just shift the original coordinate by 10
```

srwopschall replied:
Fixed.

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -645,6 +761,118 @@ bool PlanePlaneIntersection( const RealT x1, const RealT y1, const RealT z1, con
+ *       passed to this routine
+ *
+ */
+TRIBOL_HOST_DEVICE void Point3DTo2D( const RealT x, const RealT y, const RealT z, const RealT nx, const RealT ny,
```
Can't you just call the previous function with 1 point?  If a separate function is needed, maybe have the implementation just call the previous function to get rid of duplication.

srwopschall replied:
Good point. I fixed this and renamed the previous function to not be "plane" specific, just a collection of points.

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -645,6 +761,118 @@ bool PlanePlaneIntersection( const RealT x1, const RealT y1, const RealT z1, con
+ *
+ */
+TRIBOL_HOST_DEVICE void Point3DTo2D( const RealT x, const RealT y, const RealT z, const RealT nx, const RealT ny,
+                                     const RealT nz, const RealT cx, const RealT cy, const RealT cz, RealT& x_loc,
```
Though they don't really hurt, a lot of the const's aren't needed since you're copying their values.  From the perspective of the caller of the function, you don't really care whether the function changes the values or not, since they're copies, so it might be a bit cleaner to remove them.

srwopschall replied:
this routine has already been modified.

ebchin commented on src/tribol/search/InterfacePairFinder.cpp:
```diff
@@ -133,6 +135,133 @@ TRIBOL_HOST_DEVICE bool geomFilter( IndexT element_id1, IndexT element_id2, cons
+
+    // project each face's nodes to its average face plane to ensure
+    // planar faces
+    RealT x1_prime[max_nodes_per_face];
```
This projection doesn't work for mortar; you have to project non-projected coordinates to the other face.

srwopschall replied:
This is just for the purposes of filtering (check 5 inside geomFilter()). I'm not actually computing any mortar specific quantities. I suppose it could be the case that warped faces as projected to the nonmortar face may or may not have an overlap whereas warped faces projected to their average plane then projected to the nonmortar face would have the opposite? Maybe? But I don't want method specific approaches in geomFilter(). I don't even think we know the method at that level. 

If you look at `MortarPlanePair::checkFacePair()` we retain the concept of what I'm calling `prime` coordinates, which are the `planar` coordinates, but for mortar they are set to the current configuration coordinates, even if the faces are warped. This does go against the idea of making warped faces planar, but stays consistent with the process established across contact plane methods and a big note is added to make this clear.

srwopschall replied:
This has been updated in a modification to check #5 pending your review.

ebchin commented on src/tribol/search/InterfacePairFinder.cpp:
```diff
@@ -75,7 +76,8 @@ TRIBOL_HOST_DEVICE bool geomFilter( IndexT element_id1, IndexT element_id2, cons
 
   /// CHECK #4 (3D): Perform radius check, which involves seeing if
```
Looks like the check is performed in 2D also.  Maybe get rid of the `(3D)` label?

srwopschall replied:
Fixed.

ebchin commented on src/tribol/search/InterfacePairFinder.cpp:
```diff
@@ -133,6 +135,133 @@ TRIBOL_HOST_DEVICE bool geomFilter( IndexT element_id1, IndexT element_id2, cons
+    Plane3DTo2D( &x2_prime[0], &y2_prime[0], &z2_prime[0], fn2[0], fn2[1], fn2[2], cx2[0], cx2[1], cx2[2],
+                 mesh2.numberOfNodesPerElement(), &x2_bar_local[0], &y2_bar_local[0] );
+
+    Plane3DTo2D( &x1_bar[0], &y1_bar[0], &z1_bar[0], fn2[0], fn2[1], fn2[2], cx2[0], cx2[1], cx2[2],
```
Looks like this is repeated twice

srwopschall replied:
This has been removed in latest check 5

ebchin commented on src/tribol/search/InterfacePairFinder.cpp:
```diff
@@ -133,6 +135,133 @@ TRIBOL_HOST_DEVICE bool geomFilter( IndexT element_id1, IndexT element_id2, cons
+    RealT z2_bar[max_nodes_per_face];
+
+    // project 'prime' face nodal coordinates onto the plane ('bar' coords) defined by the OTHER face
+    ProjectPointsToPlane( &x1_prime[0], &y1_prime[0], &z1_prime[0], fn2[0], fn2[1], fn2[2], cx2[0], cx2[1], cx2[2],
```
You may need to walk me through what this exactly is doing, but I don't think this works for common plane either.  You want to project the 'prime' coordinates on to the common plane and do the overlap there.

srwopschall replied:
Let's go through this together. I'm fine with modifying it, but the point is that we don't have, nor should we have, method specific information at the geomFilter() level. Without knowing this or that plane I should project to given a specific method, I have to make a choice to implement this filter. I choose to project each face to its average plane, then project one face onto the other planar face and then check for a positive area of overlap. I guess the concern is that given that projection method do I miss contact plane pairs when in actuality a common plane projection or a mortar plane projection would pick them up? At one point I convinced myself this isn't an issue, but we can go through it. This is a major filtering step so we should get it right, but I don't want to introduce method specific info at this level.

srwopschall replied:
This has been modified per updated check #5 pending your review.

ebchin commented on src/tribol/search/InterfacePairFinder.cpp:
```diff
@@ -133,6 +135,133 @@ TRIBOL_HOST_DEVICE bool geomFilter( IndexT element_id1, IndexT element_id2, cons
+    RealT projX1[max_nodes_per_face];
+    RealT projY1[max_nodes_per_face];
+
+    ProjectEdgeNodesToSegment( mesh1, element_id1, fn2[0], fn2[1], cx2[0], cx2[1], &projX1[0], &projY1[0] );
```
I don't think this works either.  Here's a potential case that would get thrown out where there is a common plane overlap.

<img width="1062" height="464" alt="image" src="https://github.com/user-attachments/assets/f5c18855-d207-455d-a10e-2bf8861fa02c" />

srwopschall replied:
Let's go through this in conversation.

srwopschall replied:
This should no longer be an issue with the updated check #5.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -1662,6 +1921,130 @@ void Vertex2DOrderToCCW( const RealT* const x, const RealT* const y, RealT* xTem
+  RealT px_bar, py_bar, pz_bar;
+  ProjectPointToPlane( px, py, pz, nx, ny, nz, cx, cy, cz, px_bar, py_bar, pz_bar );
+
+  ComputeLocalBasis( px_bar, py_bar, pz_bar, nx, ny, nz, cx, cy, cz, e1x, e1y, e1z, e2x, e2y, e2z );
```
This looks broken if the normal is in the (1,1,1) direction.  You can do something like this if you need to find an orthogonal basis direction to the normal: https://mathoverflow.net/questions/87688/find-an-n-dimensional-vector-orthogonal-to-a-given-vector .  I'd create a different version of ComputeLocalBasis that only takes the normal as input and outputs e1 and e2.  The e1 guess direction can be optional and the centroid isn't needed.

srwopschall replied:
Let me take a look at this and rework if necessary.

srwopschall replied:
I introduced a test in tribol_comp_geom_3d.cpp using a plane with normal vector <1,1,1> that passes through the origin (0,0,0) and it appears to work fine. I agree you can do this in different ways, but I'm not seeing a failing case yet. I'll keep this unresolved so we have a chance to discuss.

srwopschall replied:
Leaving for now. Can revisit if necessary. Test for problematic case mentioned works with current routine.

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -7,9 +7,76 @@
+ *
+ * \pre the coordinates x,y,z lie on the plane defined by <nx,ny,nz> and <cx,cy,cz>
+ */
+TRIBOL_HOST_DEVICE void ComputeLocalBasis( RealT x, RealT y, RealT z, RealT nx, RealT ny, RealT nz, RealT cx, RealT cy,
```
See the comment in the .cpp file about creating a version that only takes in the normal (and optional e1 guess) and outputs e1 and e2.  If you take in an optional e1 guess, you should verify and correct its orthogonality to the normal.

srwopschall replied:
I'm leaving this routine as is. I checked for the case you mentioned and added a test and the computation of a local basis checks out.

ebchin replied:
It looks like the test does not test the broken case I mentioned.  If you call `ComputeLocalBasis(0.0, 0.0, 0.0, 1.0/sqrt(3.0), 1.0/sqrt(3.0), 1.0/sqrt(3.0), 0.0, 0.0, 0.0, ...)` I believe it will fail.  If it is required that x, y, z are not at cx, cy, cz, then the routine should state this and the `if (sqrMag < 1.0e-12)` check should be removed/turned into an error.

Alternatively, the routine should be refactored to not need this much input.  You can define an orthogonal basis from just a normal direction.

srwopschall replied:
I refactored this routine to just take a normal and output e1 and e2. I added tests in tribol_comp_geom_3d to test various configurations. The implementation appears to be more robust etc. Definitely take a look.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -20,6 +21,87 @@
+
+    // project point onto contact plane
+    RealT px, py, pz;
+    ProjectPointToPlane( x_new, y_new, z_new, nx, ny, nz, cx, cy, cz, px, py, pz );
```
This also seems broken if x_new and n are in the (1,1,1) direction.

ebchin replied:
See other comments about a different, robust implementation.

srwopschall replied:
I added a test in tribol_comp_geom_3d for this case and it appears to check out.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -848,7 +942,7 @@ TRIBOL_HOST_DEVICE FaceGeomError Intersection2DPolygon( const RealT* xA, const R
     int vertIdx[max_intersections];
     initIntArray( vertIdx, max_intersections, 0 );
-    PolyReorder( polyXTemp, polyYTemp, vertIdx, numPolyVert );
+    PolyReorder( &polyXTemp[0], &polyYTemp[0], &vertIdx[0], numPolyVert );
```
Why did this change?

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -645,6 +761,118 @@ bool PlanePlaneIntersection( const RealT x1, const RealT y1, const RealT z1, con
+ * \param [in] y to y-component coordinates of the edge's two vertices
+ * \param [in] xp x-coordinate of the point in question
+ * \param [in] yp y-coordinate of the point in question
+ *
```
Comment about what this returns

srwopschall replied:
added

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -1662,6 +1921,130 @@ void Vertex2DOrderToCCW( const RealT* const x, const RealT* const y, RealT* xTem
+//------------------------------------------------------------------------------
+TRIBOL_HOST_DEVICE int IsPointOrientedInEdge( const RealT* const x, const RealT* const y, RealT xp, RealT yp )
+{
+  int val = ( yp - y[0] ) * ( x[1] - xp ) - ( xp - x[0] ) * ( y[1] - yp );
```
```suggestion
  RealT val = ( yp - y[0] ) * ( x[1] - xp ) - ( xp - x[0] ) * ( y[1] - yp );
```

srwopschall replied:
Good catch. Fixed.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -1662,6 +1921,130 @@ void Vertex2DOrderToCCW( const RealT* const x, const RealT* const y, RealT* xTem
+TRIBOL_HOST_DEVICE int IsPointOrientedInEdge( const RealT* const x, const RealT* const y, RealT xp, RealT yp )
+{
+  int val = ( yp - y[0] ) * ( x[1] - xp ) - ( xp - x[0] ) * ( y[1] - yp );
+  if ( val == 0 ) return 0;
```
You might need a tolerance for this?

srwopschall replied:
Let me think about this. I see what you are saying. I'll leave unresolved for now.

srwopschall replied:
This routine has been removed. It is not required.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -1662,6 +1921,130 @@ void Vertex2DOrderToCCW( const RealT* const x, const RealT* const y, RealT* xTem
+
+  // add fuzz to catch nearly coincident vertices
+  RealT l = magnitude( x[1] - x[0], y[1] - y[0] );  // edge length
+  RealT fuzz = 0.01 * l;                            // add 1% fuzz
```
Won't this erroneously include points?  Should this be tied to one of the Tribol parameters?

srwopschall replied:
I added it as a function argument with default value of 0.0. We don't need to be overinclusive at the point this routine is used. Optionally, if a user wants to query whether a vertex is "close" to an edge, they can increase the fuzz_factor to a nonzero value.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -1662,6 +1921,130 @@ void Vertex2DOrderToCCW( const RealT* const x, const RealT* const y, RealT* xTem
+    ++k;
+  }
+
+  PolyInterYCentroid( num_nodes_1, projLocX1, projLocY1, num_nodes_2, x2Temp, y2Temp, isym, area, cy );
```
I noticed this function uses `FLT_MAX`, but we should use the largest value tied to RealT, i.e. `std::numeric_limits<RealT>::max()`.

srwopschall replied:
Testing with axom::numeric_limits<RealT>::max()

srwopschall replied:
seems to work with axom::numeric_limits::max()

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -361,6 +457,26 @@ FaceGeomError Intersection2DPolygonEnzyme( const RealT* xA, const RealT* yA, int
+ * \param [in] nV2 number of vertices on edge 2
+ * \param [in,out] overlapX pointer to x coordinates of overlapping segment
+ * \param [in,out] overlapY pointer to y coordinates of overlapping segment
+ * \param [in,out] area overlap area/length
```
```suggestion
 * \param [out] overlapX pointer to x coordinates of overlapping segment
 * \param [out] overlapY pointer to y coordinates of overlapping segment
 * \param [out] area overlap area/length
```

In general, check for `[in, out]`s that should be `[out]`s.  Here is doxygen documentation describing param: https://www.doxygen.nl/manual/commands.html#cmdparam

srwopschall replied:
I addressed them in this file. A more comprehensive check should be in another PR.

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -645,6 +761,118 @@ bool PlanePlaneIntersection( const RealT x1, const RealT y1, const RealT z1, con
+ * \param [in] y to y-component coordinates of the edge's two vertices
+ * \param [in] xp x-coordinate of the point in question
+ * \param [in] yp y-coordinate of the point in question
+ *
```
Add a `\pre` here stating `xp` is assumed collinear with the edge.

srwopschall replied:
Fixed.

ebchin commented on src/tribol/mesh/CouplingScheme.hpp:
```diff
@@ -14,13 +14,18 @@
 #endif
 
+// MPI includes
+#ifdef TRIBOL_USE_MPI
```
The MPI include isn't needed here.  Everything (tribol::CommT and TRIBOL_COMM_WORLD) is defined in `BasicTypes.hpp` which is already included.

srwopschall replied:
Removed. I think this was leftover trying to find that blueos mpi related bug.

ebchin commented on src/tribol/physics/CommonPlane.cpp:
```diff
@@ -272,35 +273,17 @@ int ApplyNormal<COMMON_PLANE, PENALTY>( CouplingScheme* cs )
     // get current configuration, physical coordinates of each face
-    mesh1.getFaceCoords( index1, &xf1[0] );
-    mesh2.getFaceCoords( index2, &xf2[0] );
+    plane.getFace1Coords( &xf1[0] );
+    plane.getFace2Coords( &xf2[0] );
```
```suggestion
    plane.getFace1Coords( xf1 );
    plane.getFace2Coords( xf2 );
```
These are already pointers

srwopschall replied:
This is a style/legacy thing that happens in a lot of places. I'll defer to a different PR to clean this up.

ebchin commented on src/tribol/physics/CommonPlane.cpp:
```diff
@@ -272,35 +273,17 @@ int ApplyNormal<COMMON_PLANE, PENALTY>( CouplingScheme* cs )
-        xVert[dim * j + 2] = cp3.m_polyZ[j];
-      }
-    }
+    plane.getOverlapVertices( &xVert[0] );
```
```suggestion
    plane.getOverlapVertices( xVert );
```

ebchin commented on src/tribol/physics/Mortar.cpp:
```diff
@@ -220,54 +211,30 @@ void ComputeSingleMortarGaps( CouplingScheme* cs )
-    }
+    // two faces; stored on the contact plane object
+    plane.getFace1Coords( &mortarX[0] );
+    plane.getFace2Coords( &nonmortarX[0] );
```
```suggestion
    plane.getFace1Coords( mortarX );
    plane.getFace2Coords( nonmortarX );
```

srwopschall replied:
Defer to another PR to clean up style/legacy things.

ebchin commented on src/tribol/physics/Mortar.cpp:
```diff
@@ -220,54 +211,30 @@ void ComputeSingleMortarGaps( CouplingScheme* cs )
-    algorithm::transpose<MemorySpace::Dynamic>( mortarX_barT, mortarX_bar );
-    algorithm::transpose<MemorySpace::Dynamic>( nonmortarX_barT, nonmortarX_bar );
+    // get face coordinates projected onto contact plane
+    plane.getFace1ProjectedCoords( &mortarX_bar[0] );
+    plane.getFace2ProjectedCoords( &nonmortarX_bar[0] );
```
```suggestion
    plane.getFace1ProjectedCoords( mortarX_bar );
    plane.getFace2ProjectedCoords( nonmortarX_bar );
```

ebchin replied:
There are quite a few more, so I'll stop commenting on each one.

srwopschall replied:
Defer to another PR to clean up style/legacy things.

ebchin commented on src/tribol/physics/Mortar.cpp:
```diff
@@ -675,7 +634,9 @@ int ApplyNormalEnzyme( CouplingScheme* cs )
   int size2 = mesh2.numberOfNodesPerElement();
 
-  for ( auto& plane : planes_view ) {
+  // for ( auto& plane : planes_view ) {
+  for ( int idx = 0; idx < num_active_pairs; ++idx ) {
+    auto& plane = cg_view.getMortarPlane( idx );
```
```suggestion
  for ( auto& plane : cg_view.getMortarPlanes() ) {
```

srwopschall replied:
Modified.

ebchin commented on src/tribol/physics/Mortar.cpp:
```diff
@@ -979,9 +940,9 @@ void ComputeMortarForceEnzyme( const RealT* x1, const RealT* n1, const RealT* p1
       Coords2DToPlaneEnzyme( tri_quad_pt, tri_quad_pt + 1, x0, e1, e2, tri_quad_pt_3d, 1 );
       RealT xi1[2] = { 0.0, 0.0 };
-      InvIso( tri_quad_pt_3d, x1t, x1t + size1, x1t + 2 * size1, size1, xi1 );
+      int inv_err = InvIso( tri_quad_pt_3d, x1t, x1t + size1, x1t + 2 * size1, size1, xi1 );
```
Why is this variable created?

srwopschall replied:
I removed the non-void return type.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1908 @@
+                                           zc2, inPlane );
+
+  TRIBOL_UNUSED_VAR( intersect1 );  // We don't currently use these bool variabeles
+  TRIBOL_UNUSED_VAR( intersect2 );  // but the above function calls modify some parameters
```
```suggestion
  LinePlaneIntersection( xA, yA, zA, xB, yB, zB, cx1[0], cx1[1], cx1z, fn1[0], fn1[1], fn1z, xc1, yc1,
                                           zc1, inPlane );

  LinePlaneIntersection( xA, yA, zA, xB, yB, zB, cx2[0], cx2[1], cx2z, fn2[0], fn2[1], fn2z, xc2, yc2,
                                           zc2, inPlane );
```

srwopschall replied:
Modified.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1908 @@
+}  // end CommonPlanePair::centroidGap()
+
+//------------------------------------------------------------------------------
+TRIBOL_HOST_DEVICE FaceGeomError CommonPlanePair::computeOverlap3D( const RealT* x1, const RealT* y1, const RealT* z1,
```
I didn't look at this in a lot of detail.  Let me know if you'd like me to look more closely, but I'm OK with the new routine provided we have sufficient testing.  Maybe we should consider some edge cases in tests to verify tolerances?

ebchin replied:
Good edge cases might be where there are line-plane intersections very close to the element edge.  Does the routine work right here?  Are there tolerances for this case?  Are the tolerances being respected?

srwopschall replied:
I added some point in face tests to tribol_comp_geom_3D. Ultimately a line-plane intersection calculation is performed, and for every actual intersection, a point in face calculation is performed to separate intersection points that lie outside vs. inside the face. There is no tolerancing in that check, just whether you are inside, on a face edge itself, or outside. I tested various point locations hovering around an edge and it performs as expected.

ebchin reviewed (changes requested):
Thanks @srwopschall for slogging through this.  The main two things that need to be addressed are comp geom check 5 and the orthogonal basis code.

srwopschall committed:
Addressing comments; removed unnecessary headers; changed tribol common
plane tied force mag check back to 0.

srwopschall committed:
Cleaned up duplicate definition of vertices in common plane penalty
test.

srwopschall committed:
Removed commented out and unused getter routines in CompGeom viewer
class.

srwopschall committed:
Simplified use of getters for common plane pairs in tests.

srwopschall committed:
Simplified getter for common plane pair in output routine.

srwopschall committed:
Removed commented out code in enzyme routine

srwopschall committed:
Removed unused test mesh object on 2D comp geom tests.

srwopschall committed:
Added a check on full overlap status for 2d test.

srwopschall committed:
Simplified register/update code in 2D comp geom tests per review
comment.

srwopschall committed:
Added clarifying comments to some of the common plane comp geom checks.

srwopschall committed:
Cleaned up how coordinates are set/handled in 2D comp geom tests

srwopschall committed:
Removed use of mortar/nonmortar language in 3D common plane tests

srwopschall committed:
Added area check function in 3d comp geom

srwopschall committed:
Added overlap area check to 2d comp geom checks.

srwopschall committed:
Trying to simplify absolute value diffs with new math function.

srwopschall committed:
Fixed typo

srwopschall committed:
Removed Point3DTo2D() and unified with Plane3DTo2D renamed to
Points3DTo2D.

srwopschall committed:
Minor review comments; removed unecessary point-to-edge orientation check

srwopschall committed:
Introduced 3d test for verifying the ComputeLocalBasis() routine.

srwopschall committed:
Hardened check #5 to catch edge cases.

srwopschall committed:
Simplifying use of virtual functions on ContactPlanePair class.

srwopschall committed:
Wrapped the contact plane data calculations in a public interface on
base class.

srwopschall committed:
use binning proximity instead of tied gap tolerance for tied contact.

Commit:
Added tolerance to local basis test

srwopschall committed:
Fixed diff tolerance in compute local basis test

srwopschall committed:
Plumbing and refactor progress on geomFilter().

srwopschall committed:
Further simplification of calls to geomFilter()

srwopschall committed:
Refactored check #5 and fixed bug in polyreorder; more work on
polyreorder to come

srwopschall committed:
Addressing minor review comments; changing EXPECT_LE to EXPECT_NEAR in
cg tests.

srwopschall committed:
Adding and changing visualization/plot default to VIS_NONE to prevent
unwanted vtk output in tests.

srwopschall committed:
Added 3D warped face test

srwopschall committed:
Added clarifying comments to PolyReorder(); simplified
checkPolyOverlap() to reflect use case

srwopschall committed:
Renamed PolyReorder() to PolyReorderConvex() for greater clarity

srwopschall committed:
style

srwopschall committed:
Added polygonal vertex reordering test; added checks on max overlap
nodes in those routines.

srwopschall commented:
@ebchin - I've addressed the items we spoke about in our review earlier this week. I have addressed the items that I wanted/needed to tackle, including introducing a warped test, and interrogating the PolyReorder routines. I've fixed one bug and introduced clarifying comments and some checks in those routines. I also added a PolyReorder tests in tribol_comp_geom_3d. Please take a look at the latest changes. Can you also review Check #5 and if you foresee major issues in the local basis routine and the projection routines, please let me know. Up until now, they all check out.

srwopschall committed:
Merge branch 'develop' into feature/swopscha/remove_separation_check

srwopschall committed:
Added point to plane projection test to cover review comment.

ebchin commented on src/tribol/mesh/CouplingScheme.cpp:
```diff
@@ -1847,4 +1868,131 @@ TRIBOL_HOST_DEVICE RealT CouplingScheme::Viewer::getGapTol( int fid1, int fid2 )
+        cx[i] = cx2[i]; 
+      }
+
+      //if (IsOverlapping( &x1[0], &y1[0], &z1[0], &x2[0], &y2[0], &z2[0], &nrml[0], &cx[0], num_nodes_face_1, num_nodes_face_2, dim ) ) {
+      //  return false;
+      //}
```
Can this be taken out?

ebchin replied:
After looking at the common plane version, looks like it needs to be un-commented?

srwopschall replied:
Good catch. Probably left commented out from debugging.

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -1662,6 +1945,158 @@ void Vertex2DOrderToCCW( const RealT* const x, const RealT* const y, RealT* xTem
+
+  return true;
+
+}
```
I like this routine.  Is the naming specific enough?  It only is for methods where we integrate over a (d-1)-dimensional hyperplane.

Should we add unit tests for this method specifically, or is there enough testing elsewhere for now?

srwopschall replied:
I added some 2D point in edge tests. That routine was not covered. Otherwise, I think there are enough tests elsewhere. To get a solution down the road post-CG you will have had to have gotten through this successfully.

srwopschall replied:
I renamed to `IsOverlappingOnPlane` and added to the doxygen your point about projecting onto a d-1 hyperplane

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -645,6 +761,86 @@ bool PlanePlaneIntersection( const RealT x1, const RealT y1, const RealT z1, con
+                                          RealT* projLocY1, RealT* projLocX2, RealT* projLocY2, RealT& area, const int isym );
+
+/*!
+ * \brief This routine is used to perform Check #5 to see if two faces/edges overlap 
```
If we don't change the name of the routine, we should make sure we note this is only for methods where we project onto a hyperplane.

Also, I think we should remove the "Check 5" reference and just stick to a description of what the routine does.

srwopschall replied:
Fixed.

ebchin commented on src/tests/tribol_common_plane_gap_rate.cpp:
```diff
@@ -70,8 +70,7 @@ void compareGaps( tribol::CouplingScheme const* cs, RealT gap, const RealT tol,
     // check diffs
-    RealT diff = std::abs( my_gap - gap );
-    EXPECT_LE( diff, tol );
+    EXPECT_LE( tribol::abs_val_diff( my_gap, gap ), tol );
```
EXPECT_NEAR

ebchin commented on src/tests/tribol_common_plane_penalty.cpp:
```diff
@@ -70,8 +70,7 @@ void compareGaps( tribol::CouplingScheme const* cs, RealT gap, const RealT tol,
     // check diffs
-    RealT diff = std::abs( my_gap - gap );
-    EXPECT_LE( diff, tol );
+    EXPECT_LE( tribol::abs_val_diff( my_gap, gap ), tol );
```
EXPECT_NEAR

ebchin commented on src/tests/tribol_common_plane_gap_rate.cpp:
```diff
@@ -130,8 +129,7 @@ void checkPressures( tribol::CouplingScheme const* cs, RealT pressure, const Rea
     // check diffs
-    RealT press_diff = std::abs( my_pressure - pressure );
-    EXPECT_LE( press_diff, tol );
+    EXPECT_LE( tribol::abs_val_diff( my_pressure, pressure ), tol );
```
EXPECT_NEAR.  Maybe do a find/replace to update all of these?

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+{
+  // outer loop over faces, inner loop over nodes/segments and determine
+  // how many 1) line-plane intersections there are (there should at most be
+  // two for interesection polygons or zero for fully separated or fully
```
```suggestion
  // two for intersection polygons or zero for fully separated or fully
```

srwopschall replied:
Fixed.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+  // outer loop over faces, inner loop over nodes/segments and determine
+  // how many 1) line-plane intersections there are (there should at most be
+  // two for interesection polygons or zero for fully separated or fully
+  // interpenetrated faces) and then 2) number of nodes one the current face
```
```suggestion
  // interpenetrated faces) and then 2) number of nodes on the current face
```

srwopschall replied:
Fixed.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+    {
+      // determine local segment vertex ids
+      int ja = j;
+      int jb = ( j == ( mesh[i]->numberOfNodesPerElement() - 1 ) ) ? 0 : ( j + 1 );
```
```suggestion
      int jb = ( j + 1 ) % mesh[i]->numberOfNodesPerElement();
```

ebchin replied:
You can also use this to get rid of the if block at L1178

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+        bool inter = LinePlaneIntersection( xa, ya, za, xb, yb, zb, cx[0], cx[1], cx[2], fn[0], fn[1], fn[2],
+                                            xInter[2 * i + k], yInter[2 * i + k], zInter[2 * i + k], inPlane );
+
+        if ( inter ) {
```
What if the intersection is at the boundary of a segment (i.e. a vertex)?  Does the intersection point get counted zero, one, or two times?

srwopschall replied:
If the edge of one face intersects the other at a vertex, then it is also likely that an edge from the other face intersects the first at, say, an edge. These intersection points will be registered as one per face wrt to the other face. Then, we check to see if the intersection point is interior, which these two should be. If one edge from each face intersect at the same identical point in space, then that point will be counted twice (one per face), but then later in the algorithm very short edge lengths of the overlapping polygon, which will also occur for coincident vertices, is checked and coincident/duplicate overlap vertices are collapsed/merged. Not sure if that answered your question.

srwopschall replied:
As a note, I added various line-plane intersection tests as I believe this was not being tested.

ebchin replied:
Wouldn't the (k < 2) check fail in that case on L1151?  Also, the error on L1141 would be triggered.

ebchin replied:
To be explicit, if both of the edge intersections are counted at the vertex, then k == 3 (or possibly k == 4) is the value that will be computed.

srwopschall replied:
Let's look at this together, but k is reset for each face. So while one face has two interior intersection points, and the other can have up to two as well, the counter is per face. The count is then stored later in a face specific array where I can see the number of intersection points (i.e. k) for each face independently. That is, they aren't added.

ebchin replied:
k == 3 or k == 4 would be the count for one face though, so the checks still look like they would fail.

ebchin replied:
<img width="790" height="566" alt="image" src="https://github.com/user-attachments/assets/8a0c8c3b-e6ec-49e5-9e0e-282ec9f0ac33" />

This is an example I had in mind.

ebchin replied:
The intersection goes through 2 vertices in the figure above, so each edge would be counted twice. k would equal 4.

srwopschall replied:
Good catch. So one face has line-plane intersections with the other face at precisely two vertices. That is, those two vertices lie exactly on the plane defined by the opposing face? I see now that four lines would register a line-plane intersection, but there are only two distinct intersection points. I'll add a test and fix up the routine.

srwopschall replied:
The code has been updated to handle this case. Namely, duplicate intersection points that may arise from this configuration are checked and the counter (k) not incremented. So k shouldn't be > 2. If it is, this is an assumption/design breaking case and the code errors out and prints the offending face coords.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+  // 2) one face has two intersection points that lie INSIDE the other face and the
+  //    other face has zero intersection points that lie INSIDE its opposing face
+  // 3) each face has two intersection points that lie INSIDE the other face; this occurs when
+  //    one face's two line-plane intersections occur at edge segments of the other face (and vice versa).
```
I don't think these are all the possible cases.

ebchin replied:
<img width="639" height="622" alt="image" src="https://github.com/user-attachments/assets/aec2b6d1-eb18-4f1e-bfd3-767c7213ad8f" />

Consider something like this: it would have 2 intersections inside on 1 plane and 1 on the other plane.  I think the check you want is total intersections inside >= 2?

srwopschall replied:
Excellent catch. I needed a second set of eyes on this. I added a test for a similar configuration that has the same intersection point relationship. One face has two intersection points, both interior but one intersecting the edge of the other face. The other face has one intersection point, and interior, but intersecting an edge of the first face. This was in fact being missed. The code has been updated to catch this.

srwopschall replied:
This case now has a test.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+    m_fullOverlap = false;
+  }
+
+  // reset degenerate intersections back to full overlap for robustness
```
We probably want to error or warning here on debug so we can look at these cases closer.  I don't think there is a case that would necessitate this (setting it explicitly to true).

srwopschall replied:
I removed this. There actually isn't a good reason to switch a degenerated interpen face-pair to a full overlap. If there is, this is not a case we understand well right now and to handle that is outside the scope of this PR. If the interpenetrating portion of one or both faces degenerates to two vertices, we throw it out. If this arises because the two faces are non-convex, for instance, then we will handle it at the convex/non-convex level, not trying to parse why the interpen portion of the face degenerated.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+  RealT x1_to_project[max_dim], y1_to_project[max_dim];
+  RealT x2_to_project[max_dim], y2_to_project[max_dim];
+
+  // project unique intersection point to the contact plane.
+  // The contact plane should have been properly located prior
+  // to this subroutine, in which case the intersection point lies
+  // on the contact plane (segment). We can still do this projection
+  // to be safe and the routine will handle a point that is already
+  // on the plane
+  // RealT xInterProj, yInterProj;
+  // ProjectPointToSegment( xInter, yInter, m_nX, m_nY, m_cX, m_cY, xInterProj, yInterProj );
```
Remove

ebchin replied:
If anything, we can do an error here, since we'd probably want to interrogate the routine when this isn't the case.

srwopschall replied:
i think for now this can be removed. This was left over from the previous implementation. This stuff about having things properly located is less of an issue I believe.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+    int interId2 = -1;
+    int k1 = 0;
+    int k2 = 0;
+    for ( int i = 0; i < m1.numberOfNodesPerElement(); ++i ) {
```
Can we do this one time?  Since for 2D, we only have 2 nodes/element and we only need to check one side?  If it's not interpenetrated, then the other side has to be interpenetrated.

srwopschall replied:
Yes this could be done differently. I'm just going to leave it now in the interest of getting this PR in and not breaking something at this point.

ebchin commented on src/tribol/geom/CompGeom.cpp:
```diff
@@ -0,0 +1,1864 @@
+    }  // end loop over nodes
+
+    // Debug check the number of interpenetrating vertices
+    if ( k1 > 1 || k2 > 1 ) {
```
Is this check needed?  I don't think it's possible to ever trigger it, provided the code above is correct.  But this should be verified via CI testing.

srwopschall replied:
I'm going to leave this in for now. Technically, and especially since we now know we are dealing with only planar faces, this should never get triggered, but there are assumptions baked in that you can only have 2 interpenetrating vertices. I left it in just in case there is some weird anomaly or numerical situation in which this is not the case. I agree, however, seems very unlikely.

srwopschall replied:
I changed the code to error out here. We shouldn't be here, but just in case. Similar to 3D I print offending edge coords.

srwopschall committed:
Reworking ComputeLocalBasis(); added more tests

srwopschall committed:
Removed old local basis routine; now using new refactored routine

srwopschall committed:
Changed enum FaceGeomError to FaceGeomException to better reflect how it
is used.

srwopschall committed:
Tightened up force diff tolerance on common plane gpu test.

srwopschall committed:
Fixed bug in IsOverlapping() 2D check to include BOTH 1 in 2 AND 2 in 1.

srwopschall committed:
Added fuzz_factor input arg to IsPointInEdge()

srwopschall committed:
Changed FLT_MAX to max of RealT in simpler overlap routine

srwopschall committed:
Fixed some [in,out] to [out]

srwopschall committed:
Testing axom::numeric_limits in simpler overlap routine

srwopschall committed:
Various minor review comments.

srwopschall committed:
Added point-in-face tests to test edge cases with vertex hovering around
edges of face

srwopschall committed:
Uncommented IsOverlapping for mortar method.

srwopschall committed:
Added point in edge test with various edge case configurations.

srwopschall committed:
Modified naming of IsOverlapping() to be more specific

srwopschall committed:
Changed EXPECT_LT to EXPECT_NEAR where noted in review.

srwopschall committed:
Added 3D test for various line-plane intersection configurations

srwopschall committed:
Added missed case switching from full to interpen overlap

srwopschall committed:
Added check for number of interpen overlap vertices in 3D cg tests

srwopschall committed:
Removed old code in 2d overlap routine

srwopschall committed:
Updated Area2DPolygon to shoelace alg to handle non-convex overlaps;
more 3D tests for edge cases; updated mfem_common_plane max_disp
baseline.

srwopschall committed:
Added more comments about the duplicate interpen vertex check.

srwopschall committed:
Printing offending common plane face coords prior to erroring out for
too many line-plane intersections

srwopschall committed:
Minor comment formatting.

srwopschall committed:
Added 2D test with x-like configuration with very small epsilon
rotation.

srwopschall committed:
Added more 2D and 3D edge case tests

srwopschall committed:
Changed 2D CG to error out if there is more than one intersection point.

srwopschall committed:
Removed switch to full overlap for degenerated interpenetrating face
portions.

srwopschall committed:
Added convexity routine; incorporated convexity into common plane CG;
modified nonconvex test accordingly

srwopschall committed:
formatting

srwopschall committed:
Loosened mfem common plane tolerance based on observed testing/platform
differences.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -545,6 +599,316 @@ TEST_F( CompGeomTest, common_plane_interpen_check_3 )
+  auto& plane = couplingScheme->getCompGeom().getCommonPlane( 0 );
+  EXPECT_EQ( plane.m_inContact, true );
+}
+
```
Nice!

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -545,6 +599,316 @@ TEST_F( CompGeomTest, common_plane_interpen_check_3 )
+  y1[1] = 0.0;
+
+  x2[0] = 0.;
+  x2[1] = 0.99999999999;
```
maybe use an explicit epsilon w/exponential notation, so it's clearer how the vertices are moving

srwopschall replied:
Fixed.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -545,6 +599,316 @@ TEST_F( CompGeomTest, common_plane_interpen_check_3 )
+  y1[0] = 0.0;
+  y1[1] = 0.0;
+
+  RealT epsilon = 1.e-10;
```
like this test.  let's try a smaller epsilon, i.e. less than the tolerances in the cg routines, and see what happens.  we could potentially loop over a range of epsilon

srwopschall replied:
I can use 1.e-12 and still get the desired behavior. 1.e-15 won't register interpen when interpen is happening. It just sees faces or vertices as right on top of one another. I think this is reasonable. I lowered the appropriate epsilons to 1.e-12 and made a comment.

ebchin commented on src/tests/tribol_comp_geom_2d.cpp:
```diff
@@ -704,6 +1067,60 @@ TEST_F( CompGeomTest, 2d_projections_2 )
+  EXPECT_EQ( tribol::IsPointInEdge( x, y, xp, yp ), false );
+
+  // point just barely outside second vertex
+  xp = 1.5000000000001;
```
an explicit epsilon (in exp notation) would make this clearer

srwopschall replied:
Fixed.

ebchin commented on src/tests/tribol_comp_geom_3d.cpp:
```diff
@@ -714,26 +723,592 @@ TEST_F( CompGeomTest, common_plane_single_element_interpen_check_3_warped )
+    z[i] = z_rot;  //+ z_shift;
+  }
+
+  // Debug print vertices
```
Can these lines be removed?  Have you done this verification for all of the tests?  Just want to make sure the test is seeing the geometry we think it should be seeing.

srwopschall replied:
I removed them. Yes, for all tests I plotted the face-pair to make sure I have what I think I have.

ebchin commented on src/tests/tribol_comp_geom_3d.cpp:
```diff
@@ -817,36 +1380,45 @@ TEST_F( CompGeomTest, common_plane_single_element_interpen_check_4 )
+    z[i] = z_rot;
   }
 
+  // Debug print vertices
```
Remove?

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -21,57 +21,39 @@
+
+  // now compute the second basis vector as n x e1
+  crossProd( nx, ny, nz, e1x, e1y, e1z, e2x, e2y, e2z );
+}
```
Very nice

ebchin commented on src/tribol/geom/GeomUtilities.cpp:
```diff
@@ -2076,27 +2041,67 @@ TRIBOL_HOST_DEVICE bool IsOverlapping( const RealT* const x1, const RealT* const
+
+    RealT cross = ax * by - ay * bx;
+
+    if ( cross > 0 ) {
```
What if cross == 0?

srwopschall replied:
That means two segments sharing a node are actually colinear. This does not break convexity, but per this algorithm it does not indicate the turning direction. Here, we only need to know if all cross products are either negative or positive.

ebchin commented on src/tribol/geom/GeomUtilities.hpp:
```diff
@@ -835,11 +841,28 @@ TRIBOL_HOST_DEVICE void CheckPolyOverlap( const int num_nodes_1, const int num_n
  * \return true if the faces/edges overlap; otherwise false
+ *
+ * \note this routine is only to check whether two edges or faces overlap as projected onto a
+ *       common or intermediate contact plane (i.e. a (d-1) - hyperplane.
```
```suggestion
 *       common or intermediate contact plane (i.e. a (d-1)-dimensional hyperplane).
```

srwopschall replied:
Fixed.

ebchin commented on src/tribol/utils/Math.cpp:
```diff
@@ -120,10 +120,7 @@ void bubble_sort( int* const array, const int n )
-{
-  return std::abs(val1 - val2);
-}
+RealT abs_val_diff( RealT val1, RealT val2 ) { return std::abs( val1 - val2 ); }
```
Is this still used?  Can remove if not.

srwopschall replied:
It is used in some tests. Cleaned things up a little. I'll just leave it.

ebchin commented on src/tribol/utils/Math.hpp:
```diff
@@ -70,8 +70,8 @@ void bubble_sort( int* const array,  ///< [in] Input array of integers
 /// compute the absolute value of the difference between two values
-RealT abs_val_diff( RealT val1, ///< [in] first value
-                    RealT val2  ///< [in] second value
+RealT abs_val_diff( RealT val1,  ///< [in] first value
```
Same with this one.

srwopschall committed:
Merged with develop; fixed point in triangle checks

srwopschall committed:
Addressing review comments

srwopschall committed:
Fixed enzyme mortar function

srwopschall committed:
style

srwopschall committed:
Added host config for enzyme build

srwopschall committed:
Merge branch 'develop' into feature/swopscha/remove_separation_check

ebchin committed:
add num coords to get face coords routines

srwopschall committed:
Removed duplicate note in release notes

srwopschall committed:
style

srwopschall committed:
Reverted back to FLT_MAX where appropriate

srwopschall committed:
Added check for interpen intersection point is node

srwopschall committed:
style

srwopschall committed:
Test cleanup

srwopschall committed:
Remove debug code

srwopschall committed:
Added notes to the IsConvex() routine for clarity.

srwopschall committed:
Minor update to release notes.
